tests/e2e/test_subshell.py

```diff
@@ -15,3 +15,85 @@ def test_subshell_with_pipeline(shell_session):
 
     result = get_command_output(shell_session.before)
     assert "Hello$" == result
+
+
+def recursive(shell_session, i, INITIAL_SHLVL):
```

Comment:
It would be better to name the function `check_shlvl_increment_recursively` instead of `recursive`.
In Python, capital case represents global or constant variables. So `initial_shlvl` is better than `INITIAL_SHLVL` for this parameter.